src/system_extra.rs

```diff
@@ -0,0 +1,105 @@
+//! Extra system utilities.
+//!
+//! This modules contains an extension trait for the System trait which adds useful transformation
+//! functions.
+
+use ecs::prelude::{Read, System};
+use shred::{RunningTime, SystemData};
+
+/// Extra functionality associated systems.
+pub trait SystemExtra {
+    /// Build a system that will do nothing unless the `V` resource the specified value.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use amethyst::{
+    ///     ecs::{System, Write},
+    ///     shred::DispatcherBuilder,
+    ///     prelude::*,
+    /// };
+    ///
+    /// #[derive(PartialEq, Eq)]
+    /// enum State {
+    ///     Other,
+    ///     Example,
+    /// }
+    ///
+    /// impl Default for State {
+    ///     fn default() -> Self {
+    ///         State::Other
+    ///     }
+    /// }
+    ///
+    /// struct Example(u32);
+    ///
+    /// impl<'s> System<'s> for Example {
+    ///     type SystemData = Write<'s, u32>;
+    ///
+    ///     fn run(&mut self, mut number: Self::SystemData) {
+    ///         *number = self.0;
+    ///     }
+    /// }
+    ///
+    /// let mut world = World::new();
+    ///
+    /// let mut dispatcher = DispatcherBuilder::default()
+    ///     .with(Example(42), "set_number", &[])
+    ///     .with(Example(84).enabled_on(State::Example), "set_number_2", &[])
```

Comment:
Thinking about this, this only allows intersection, but no unions. I'm wondering if we can allow both.

Reply:
Interesting!

I messed around with two different approaches:
* Store a collection to check if the enum is a member of that collection (e.g. `HashSet<State>`).
* Store the value as-is right now, but implement more complex testing associated with the system (e.g. `OneOf<HashSet<State>>`).

I landed on the second approach, because it ended up being significantly cleaner. More generic parameters could be elided instead of explicitly specified.

Here it is:
https://github.com/udoprog/amethyst/blob/enabled-system-predicates/src/system_extra.rs#L182

What do you think?

Reply:
Part of me likes the pattern, but at the same time I'm unsure if this is actually helpful. I think if we show in the docs how to do more complex things (with concrete types), it helps them more than a generic API. So my suggestion to go forward with this would be making this very simple, just a `Pausable` system without any values and to show in the docs and book how to do more complex checks.

Reply:
Oh, I think I understand now, you were expecting this to be used only for pausing a _single_ set of systems?

I envisioned this to be used with having a set of systems _per state_ that should be running, not only for the pause functionality.

As an example, you have a set of systems running for:
* The game menu (UI + Menu Inputs).
* The main game (Game UI + Game Logic + Game Inputs).
* The game is paused (Pause Inputs).